File: python/src/bw_stt/transcript.py
"""Helpers that turn Segment events into display text."""

from __future__ import annotations

from dataclasses import dataclass

from .events import Segment

__all__ = ["DisplayWord", "TranscriptAssembler", "WordAssembler"]


class TranscriptAssembler:
    """Rebuild the full transcript by plain concatenation of segment text.

    For a split-stereo session (``multichannel=True``), use one assembler
    per channel.
    """

    def __init__(self) -> None:
        self._text = ""

    @property
    def text(self) -> str:
        """The transcript assembled so far."""
        return self._text

    def push(self, segment: Segment) -> str:
        """Append one segment and return the current full transcript."""
        self._text += segment.text
        return self._text


@dataclass(frozen=True)
class DisplayWord:
    """One merged display word spanning its first piece's start to its last piece's end."""

    text: str
    start: float
    end: float


class WordAssembler:
    """Merge segment word pieces into display words for live rendering.

    A segment whose text starts with a space begins a new word; one without
    a leading space continues the previous word, so its first word entry
    extends the last display word in place. For a split-stereo session
    (``multichannel=True``), use one assembler per channel.
    """

    def __init__(self) -> None:
        self._words: list[DisplayWord] = []

    @property
    def words(self) -> list[DisplayWord]:
        """The display words merged so far."""
        return list(self._words)

    def push(self, segment: Segment) -> list[DisplayWord]:
        """Merge one segment and return the current display words."""
        continues = bool(self._words) and not segment.text.startswith(" ")
        for index, word in enumerate(segment.words):
            if index == 0 and continues:
                last = self._words[-1]
                self._words[-1] = DisplayWord(last.text + word.word, last.start, word.end)
            else:
                self._words.append(DisplayWord(word.word, word.start, word.end))
        return list(self._words)

File: typescript/src/transcript.ts
import type { Segment } from "./events";

/**
 * Builds the running transcript by verbatim concatenation of segment text.
 * Segments carry their own leading spaces; never insert separators.
 */
export class TranscriptAssembler {
  private buffer = "";

  /** Appends one segment and returns the full transcript so far. */
  push(segment: Pick<Segment, "text">): string {
    this.buffer += segment.text;
    return this.buffer;
  }

  get text(): string {
    return this.buffer;
  }
}

/** A display word merged from one or more subword pieces. */
export interface DisplayWord {
  text: string;
  /** Seconds; spans the first piece's start. */
  start: number;
  /** Seconds; extends to the latest piece's end. */
  end: number;
}

/**
 * Builds display words for live rendering. A segment whose text starts with a
 * space begins a new word; one without a leading space continues the previous
 * word in place (subword pieces), extending its text and end time. Use one
 * assembler per channel in multichannel sessions.
 */
export class WordAssembler {
  private readonly list: DisplayWord[] = [];

  /**
   * Merges one segment and returns the assembler's live word list. The same
   * array instance is returned on every call and grows in place.
   */
  push(segment: Pick<Segment, "text" | "words">): DisplayWord[] {
    const continuesPrevious = !segment.text.startsWith(" ") && this.list.length > 0;
    segment.words.forEach((word, index) => {
      if (index === 0 && continuesPrevious) {
        const previous = this.list[this.list.length - 1]!;
        previous.text += word.word;
        previous.end = word.end;
      } else {
        this.list.push({ text: word.word, start: word.start, end: word.end });
      }
    });
    return this.list;
  }

  get words(): readonly DisplayWord[] {
    return this.list;
  }

  /** The current words joined with single spaces. */
  get text(): string {
    return this.list.map((word) => word.text).join(" ");
  }
}
